radixdlt/src/main/java/com/radixdlt/consensus/QuorumCertificate.java

```diff
@@ -84,10 +99,43 @@ public VoteData getVoteData() {
 		return voteData;
 	}
 
-	public ECDSASignatures getSignatures() {
+	public TimestampedECDSASignatures getSignatures() {
 		return signatures;
 	}
 
+	public long quorumTime(ValidatorSet validatorSet) {
+		// Note that signatures are not rechecked here.
+		// They are expected to be checked before the QC is formed.
+		// No effort is made to ensure the QC is valid for this validator set.
+		UInt256 totalPower = UInt256.ZERO;
+		ImmutableMap<ECPublicKey, Validator> validators = validatorSet.validatorsByKey();
+		List<Pair<Long, UInt256>> weightedTimes = Lists.newArrayList();
+		for (Map.Entry<ECPublicKey, Pair<Long, ECDSASignature>> e : this.signatures.getSignatures().entrySet()) {
+			ECPublicKey thisKey = e.getKey();
+			Validator v = validators.get(thisKey);
+			if (v != null) {
+				UInt256 power = v.getPower();
+				totalPower = totalPower.add(power);
+				weightedTimes.add(Pair.of(e.getValue().getFirst(), power));
+			}
+		}
+		if (totalPower.isZero()) {
+			// You probably used the wrong validator set
+			throw new IllegalStateException("Zero validator power for this QC");
+		}
+		UInt256 median = totalPower.shiftRight(); // Divide by 2
+		// Sort ascending by timestamp
+		weightedTimes.sort(Comparator.comparing(Pair::getFirst));
+		for (Pair<Long, UInt256> w : weightedTimes) {
+			UInt256 weight = w.getSecond();
+			if (median.compareTo(weight) < 0) {
+				return w.getFirst();
+			}
+			median = median.subtract(weight);
```

Comment:
So we're trying to get the median time but by weight?
Wouldn't the median weight have to be averaged at "the edges" to be a proper median?

Reply:
Hmm.  I thought about this, but I think the timestamp really should be one of the submitted timestamps, rather than an average and the chances of the median "falling in the gap" seems pretty low.